File: src/main/java/com/github/simiacryptus/aicoder/util/BlockComment.java
package com.github.simiacryptus.aicoder.util;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.stream.Collectors;

public class BlockComment extends IndentedText {
    public static class Factory implements TextBlockFactory<BlockComment> {
        public final String linePrefix;
        public final String blockPrefix;
        public final String blockSuffix;

        public Factory(String blockPrefix, String linePrefix, String blockSuffix) {
            this.linePrefix = linePrefix;
            this.blockPrefix = blockPrefix;
            this.blockSuffix = blockSuffix;
        }

        @Override
        public BlockComment fromString(String text) {
            text = StringTools.stripSuffix(StringTools.trimSuffix(text.replace("\t", TAB_REPLACEMENT)), blockSuffix.trim());
            @NotNull CharSequence indent = StringTools.getWhitespacePrefix(text.split(DELIMITER));
            return new BlockComment(blockPrefix, linePrefix, blockSuffix, indent, Arrays.stream(text.split(DELIMITER))
                    .map(s -> StringTools.stripPrefix(s, indent))
                    .map(StringTools::trimPrefix)
                    .map(s -> StringTools.stripPrefix(s, blockPrefix.trim()))
                    .map(s -> StringTools.stripPrefix(s, linePrefix.trim()))
                    .toArray(CharSequence[]::new));
        }

        @Override
        public boolean looksLike(String text) {
            return text.trim().startsWith(blockPrefix) && text.trim().endsWith(blockSuffix);
        }
    }

    public final CharSequence linePrefix;
    public final CharSequence blockPrefix;
    public final CharSequence blockSuffix;


    public BlockComment(CharSequence blockPrefix, CharSequence linePrefix, CharSequence blockSuffix, CharSequence indent, CharSequence... textBlock) {
        super(indent, textBlock);
        this.linePrefix = linePrefix;
        this.blockPrefix = blockPrefix;
        this.blockSuffix = blockSuffix;
    }

    @Override
    public @NotNull String toString() {
        CharSequence indent = getIndent();
        CharSequence delimiter = DELIMITER + indent;
        CharSequence joined = Arrays.stream(rawString()).map(x->linePrefix + " " + x).collect(Collectors.joining(delimiter));
        return blockPrefix.toString() + delimiter + joined + delimiter + blockSuffix;
    }

    public @NotNull BlockComment withIndent(CharSequence indent) {
        return new BlockComment(blockPrefix, linePrefix, blockSuffix, indent, textBlock);
    }
}

File: src/main/kotlin/com/github/simiacryptus/aicoder/util/BlockComment.kt
@file:Suppress("NAME_SHADOWING")

package com.github.simiacryptus.aicoder.util

import com.simiacryptus.util.StringUtil
import java.util.*
import java.util.stream.Collectors

class BlockComment(
    val blockPrefix: CharSequence,
    val linePrefix: CharSequence,
    val blockSuffix: CharSequence,
    indent: CharSequence,
    vararg textBlock: CharSequence
) :
    IndentedText(indent, *textBlock) {
    class Factory(val blockPrefix: String, val linePrefix: String, val blockSuffix: String) :
        TextBlockFactory<BlockComment?> {
        override fun fromString(text: String?): BlockComment {
            var text = text!!
            text = StringUtil.stripSuffix(
                StringUtil.trimSuffix(text.replace("\t", TAB_REPLACEMENT.toString(), false)),
                blockSuffix.trim { it <= ' ' })
            val indent = StringUtil.getWhitespacePrefix(*text.split(DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray())
            return BlockComment(blockPrefix,
                linePrefix,
                blockSuffix,
                indent,
                *(Arrays.stream(text.split(DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
                    .map { s: CharSequence? -> StringUtil.stripPrefix(s!!, indent) }
                    .map { text: CharSequence? -> StringUtil.trimPrefix(text!!) }
                    .map { s: CharSequence? -> StringUtil.stripPrefix(s!!, blockPrefix.trim { it <= ' ' }) }
                    .map { s: CharSequence? -> StringUtil.stripPrefix(s!!, linePrefix.trim { it <= ' ' }) }
                    .collect(Collectors.toList()).toTypedArray()))
        }

        override fun looksLike(text: String?): Boolean {
            return text!!.trim { it <= ' ' }.startsWith(blockPrefix) && text.trim { it <= ' ' }.endsWith(blockSuffix)
        }
    }

    override fun toString(): String {
        val indent = this.indent
        val delimiter: CharSequence = DELIMITER + indent
        val joined: CharSequence = Arrays.stream(rawString()).map { x: CharSequence -> "$linePrefix $x" }
            .collect(Collectors.joining(delimiter))
        return blockPrefix.toString() + delimiter + joined + delimiter + blockSuffix
    }

    override fun withIndent(indent: CharSequence?): BlockComment {
        return BlockComment(blockPrefix, linePrefix, blockSuffix, indent!!, *lines)
    }
}
